### PR TITLE
hmm + gmm: Add timing information to `fit` output

### PR DESCRIPTION
**Pull request by mrshu**
* Add timing information to verbose output the `fit` function of GMM and
  HMM, as described in #175.

Signed-off-by: mr.Shu <mr@shu.io>

### CODE REVIEW

**Comment by mrshu** [comment]
@jmschrei Sorry for so many commits -- I'll gladly squash them.

It turns out that one cannot really rely on the presence of `clock_gettime` on all of the considered platforms. In hindsight, I should have done it this way since the beginning -- getting the timing info does not necessarily have to be optimized in this case.

File: pomegranate/gmm.pyx [hunk 1]
```diff
@@ -170,8 +170,8 @@ cdef class GeneralMixtureModel(BayesModel):
 			else:
 				improvement = log_probability_sum - last_log_probability_sum
 
-				time_spent = time_in_epoch_sec() - epoch_start_time
-				msg = "Improvement: {} in {:.2f}"
+				time_spent = time.time() - epoch_start_time
+				msg = "Improvement: {} in {:.2f}s"
```
Comment (jmschrei) [inline pomegranate/gmm.pyx]: Might as well just make this

```
if verbose:
    print "Improvement: {} in {:.2f}".format(improvement, time_spent)
``

**Comment by jmschrei** [comment]
Getting the current clock time is so inconsequential compared to the main computations that I don't think you should bother. Even if it took 100 ms, no one would really notice, and usually it takes less than 1ms.

**Comment by mrshu** [comment]
Thanks for your assessment @jmschrei, I agree with that.

The checks have passed, so this should be safe to merge now.

File: pomegranate/gmm.pyx [hunk 2]
```diff
@@ -167,17 +170,20 @@ cdef class GeneralMixtureModel(BayesModel):
 			else:
 				improvement = log_probability_sum - last_log_probability_sum
 
+				time_spent = time.time() - epoch_start_time
 				if verbose:
-					print("Improvement: {}".format(improvement))
+					print("Improvement: {} in {:.2f}s".format(improvement, time_spent))
```
Comment (mrshu) [inline pomegranate/gmm.pyx]: @jmschrei I am not sure if you meant it this way in your comment (I believe the `s` at the end was omitted), but the reason why I tried to split this into two lines was that currently the line is a bit too long when you set your `tabwidth` to 8.

However, since there are even longer lines in this file, this is acceptable in my mind too.

File: pomegranate/hmm.pyx [hunk 3]
```diff
@@ -2486,8 +2490,10 @@ cdef class HiddenMarkovModel(GraphModel):
                     initial_log_probability_sum = log_probability_sum
                 else:
                     improvement = log_probability_sum - last_log_probability_sum
+                    time_spent = time.time() - epoch_start_time
                     if verbose:
-                        print("Training improvement: {}".format(improvement))
+                        print("Training improvement: {} in {:.2f}s".format(improvement,
+                                                                           time_spent))
```
Comment (mrshu) [inline pomegranate/hmm.pyx]: @jmschrei Just a quick note with regards to #156 -- the alignment of `improvement` and `time_spent` here is possible outright due to the use of spaces. If spaces were not used, the first part of the line (up until `print`) would have to be indented with tabs and from then on spaces could be used. This would make it quite difficult to put together.

**Comment by mrshu** [comment]
@jmschrei The checks have passed -- I have contemplated adding some more tests for this. In the end, it does not seem worth it as their value would be quite negligible. I'll try to add more of them when implementing the other parts of #175.

Please let me know if there Is anything else you think should be done before merging this.

Thanks!

File: pomegranate/gmm.pyx (hunk 2 above)
Comment (jmschrei) [inline pomegranate/gmm.pyx]: I suppose it doesn't really matter one way or another. This looks fine.

File: pomegranate/hmm.pyx (hunk 3 above)
Comment (jmschrei) [inline pomegranate/hmm.pyx]: Ultimately it doesn't matter too much. It is obvious what is going on to people who want to develop.

**Comment by jmschrei** [comment]
Is it possible to change the format to be as follows:

[iteration] Improvement: {improvement}, Time (s): {time}

starting with 1 as the first iteration, not 0. Thanks!

**Comment by mrshu** [comment]
@jmschrei Sure, I'll do that right away!

File: pomegranate/hmm.pyx (hunk 3 above)
Comment (mrshu) [inline pomegranate/hmm.pyx]: Agreed for sure.

File: pomegranate/gmm.pyx [hunk 4]
```diff
@@ -167,17 +170,22 @@ cdef class GeneralMixtureModel(BayesModel):
 			else:
 				improvement = log_probability_sum - last_log_probability_sum
 
+				time_spent = time.time() - epoch_start_time
+				vals = dict(iteration=iteration, improvement=improvement, time=time_spent)
+				fmt = "[{iteration}] Improvement: {improvement} Time (s): {time:.2f}"
```
Comment (mrshu) [inline pomegranate/gmm.pyx]: @jmschrei I realize you may find this to be a bit too much, but I believe that it is better to create a variable this way than trying to fit it into the `print` line below, which would then overflow.

File: pomegranate/gmm.pyx [hunk 5]
```diff
@@ -167,17 +170,22 @@ cdef class GeneralMixtureModel(BayesModel):
...
 			last_log_probability_sum = log_probability_sum
 
 		self.clear_summaries()
 
 		if verbose:
-			print("Total Improvement: {}".format(
-				last_log_probability_sum - initial_log_probability_sum))
+			total_imp = last_log_probability_sum - initial_log_probability_sum
+			total_time_spent = time.time() - training_start_time
+			print("Total Improvement: {}".format(total_imp))
```
Comment (mrshu) [inline pomegranate/gmm.pyx]: @jmschrei Do these prints look OK? I pretty much only guessed their format.

File: pomegranate/FactorGraph.pyx [hunk 6]
```diff
@@ -395,7 +395,7 @@ cdef class FactorGraph( GraphModel ):
 			"states": [json.loads(state.to_json()) for state in self.states],
 			"edges": self.edges
 		}
-		return json.dumps(model, separators=separators, indent=indent)
+		return json.dumps(model, separators=separators, indent=indent, sort_keys=True)
```
Comment (mrshu) [inline pomegranate/FactorGraph.pyx]: @jmschrei It turns out this was necessary in order for the JSON comparison test of factor graphs (https://github.com/jmschrei/pomegranate/pull/267/files#diff-6ba2f6007ae6a66d1e3941680c09a027R31) to work reliably.

Comment (jmschrei) [inline pomegranate/FactorGraph.pyx]: Great catch, I just noticed a test failed on AppVeyor.

File: pomegranate/gmm.pyx (hunk 4 above)
Comment (jmschrei) [inline pomegranate/gmm.pyx]: I don't mind merging it now and then revisiting the issue later. Better to have the feature then clean it up.

Comment (mrshu) [inline pomegranate/gmm.pyx]: Cool, thanks!

File: pomegranate/FactorGraph.pyx (hunk 6 above)
Comment (mrshu) [inline pomegranate/FactorGraph.pyx]: I am not sure if you would like me to send another patch specifically for this -- I am happy to do that, just let me know.

**Comment by mrshu** [comment]
@jmschrei From my point of view, this PR is done so please let me know how you would like me to proceed here. I see two options:

1. Just merge this PR, which will both add the timing info and fix the FactorGraph test issue
2. Create a separate PR for the FactorGraph test issue, get that merged and then merge this PR.

Whichever option you choose, I am happy to help you with.

Thanks!

**Comment by jmschrei** [comment]
I will be getting to this soon-- very busy recently! Thank you for your patience.

**Comment by mrshu** [comment]
@jmschrei No worries at all -- I know very well how that is!

In the meantime, I added timing info to `BayesClassifier`, `NaiveBayes` and also `kmeans`, as their `fit` function was pretty similar to that of `gmm` and `hmm`. I believe this would be a nice addition, but I am happy to discuss it.

**Comment by jmschrei** [comment]
Great catch, semi-supervised learning using those models is also iterative. I think those are the only ones, though. Bayes nets, Markov chains, and probability distributions just use MLE for parameter updates.

**Comment by mrshu** [comment]
@jmschrei thanks, that is indeed what I gathered from grepping around through the repository.
